agent/src/beerocks/slave/tasks/task_pool.h

```diff
@@ -0,0 +1,66 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _TASK_POOL_H_
+#define _TASK_POOL_H_
+
+#include "task.h"
+#include <unordered_map>
+
+namespace beerocks {
+
+/**
+ * @brief The TaskPool is single instance tasks container which responsible to run each
+ * task and allow passing messages and events from the Agent to the task.
+ * 
+ */
+class TaskPool {
+public:
+    /**
+     * @brief Add new task to the task pool.
+     * 
+     * @param new_task Shared pointer to the task.
+     * @return true on success, otherwise false.
+     */
+    bool add_task(std::shared_ptr<Task> new_task);
```

Comment:
I think it's safe to make this a const reference, no need to create additional copy as another one will be created when storing the task in the map:
```suggestion
    bool add_task(const std::shared_ptr<Task> &new_task);
```

Reply:
But then the reference count will not increase, and the memory will be freed at the end of the caller scope.
And the task will hold invalid memory.
Am I missing something?

Reply:
The reference count will increase once the pointer is stored in the map.
When you pass it by value, the reference count increases twice:
1. When the pointer is copied to the function argument.
2. When the pointer is stored in the map.

Once the function ends, the reference count decreases by 1.
By passing a reference, you prevent the extra increase-decrease.

Reply:
done. :)